pkg/serviceaccount/metadata/metadata.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package metadata is a separate subpackage, because many controllers
+// also import pkg/serviceaccount, but are not allowed by import-boss
+// to depend on go-restful. This way, this code stays close to other related
+// code, but doesn't add go-restful as a dependency of those controllers.
+package metadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	restful "github.com/emicklei/go-restful"
+	jose "gopkg.in/square/go-jose.v2"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+const (
+	// OpenIDConfigPath is the URL path at which the API server serves
+	// an OIDC Provider Configuration Information document, corresponding
+	// to the Kubernetes Service Account key issuer.
+	// https://openid.net/specs/openid-connect-discovery-1_0.html
+	OpenIDConfigPath = "/.well-known/openid-configuration"
+	// JWKSPath is the URL path at which the API server serves a JWKS
+	// containing the public keys that may be used to sign Kubernetes
+	// Service Account keys
+	JWKSPath = "/jwks"
+	// cacheControlMaxAge is the value (seconds) of the max-age Cache-Control header
+	cacheControlMaxAge = "max-age=86400" // 1 day
+)
+
+// IssuerMetadataServer is an HTTP server for metadata of the KSA token issuer.
+type IssuerMetadataServer struct {
+	metadataJSON []byte
+	keysetJSON   []byte
+}
+
+// issuerMetadata provides a minimal subset of OIDC provider metadata:
+// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+type issuerMetadata struct {
+	Issuer string `json:"issuer"` // REQUIRED in OIDC; meaningful to relying parties.
+	// AuthzEndpoint string   `json:"authorization_endpoint"`                // REQUIRED in OIDC; but useless to relying parties.
```

Comment:
If this is required by the RFC, we must not omit this field entirely. @mikedanese, any creative spec-compliant ideas for a value here that will not make naive clients send requests to a bad place?

Reply:
We explicitly called this out in the KEP and decided to omit until we confirm incompatibility or determine a valid placeholder value.